pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -1231,37 +1232,56 @@ private void asyncDeleteCursorWithClearDelayedMessage(String subscriptionName,
             if (ex != null) {
                 unsubscribeFuture.completeExceptionally(ex);
             } else {
-                asyncDeleteCursor(subscriptionName, unsubscribeFuture);
+                asyncDeleteCursor(persistentSubscription, unsubscribeFuture);
             }
         });
     }
 
-    private void asyncDeleteCursor(String subscriptionName, CompletableFuture<Void> unsubscribeFuture) {
```

Comment:
To avoid so many code changes due to a pre-processing, you can add an override like:

```java
    private void asyncDeleteCursor(PersistentSubscription subscription, CompletableFuture<Void> unsubscribeFuture) {
        final var subscriptionName = subscription.getName();
        if (!isCompactionSubscription(subscriptionName) || !(subscription instanceof PulsarCompactorSubscription)) {
            asyncDeleteCursor(subscriptionName, unsubscribeFuture);
            return;
        }
        ((PulsarCompactorSubscription) subscription).cleanCompactedLedger()
                /* ... */
                .thenAccept(__ -> asyncDeleteCursor(subscriptionName, unsubscribeFuture));
    }
```